### PR TITLE
feat(babel-plugin-compiler): observe string, number, and identifier bracketed fields

### LINKED ISSUE #4547: Reactivity only works for non-bracket-notation string properties
```js
export default class extends LightningElement {
  reg = 'regular!';
  ["with spaces"] = 'spaces!';
  [1337] = 'number!';
  [symbol] = 'symbol!';
}
```

In the above example, only `reg` is a reactive property. The other properties are non-reactive, because `@lwc/babel-plugin-component` does not recognize them as reactive properties and doesn't register them as such:

```js
registerDecorators(App, {
  fields: ["reg"]
});
```

[Repro showing reactivity not working correctly](https://stackblitz.com/edit/salesforce-lwc-otqenf?file=src%2Fmodules%2Fx%2Fapp%2Fapp.html,src%2Fmodules%2Fx%2Fapp%2Fapp.js&title=LWC%20playground)

This is definitely an edge case of an edge case, but maybe it's something we should support? If not, it's at least worth documenting in this GitHub issue.

<details><summary>Full repro for posterity</summary>

```html
<template>
    <strong>Regular</strong>: {regular}
    <strong>Spaces</strong>: {spaces}
    <strong>Number</strong>: {number}
    <strong>Symbol</strong>: {symbol}
</template>
```

```js
import { LightningElement } from 'lwc';

const symbol = Symbol("haha")
let count = 0
export default class App extends LightningElement {
  reg = 'regular!';
  ["with spaces"] = 'spaces!';
  [1337] = 'number!';
  [symbol] = 'symbol!';

  get regular() {
    return this.reg
  }

  get spaces() {
    return this['with spaces']
  }

  get number() {
    return this[1337]
  }

  get symbol () {
    return this[symbol]
  }

  renderedCallback() {
    if (++count) {
      setTimeout(() => {
        this.reg = 'regular 2!'
        setTimeout(() => {
          this['with spaces'] = 'spaces 2!'
          this[1337] = 'number 2!'
          this[symbol] = 'symbol 2!'
        })
      })
    }
  }
}

```

</details>

### PR DESCRIPTION
**Pull request by cardoso**
## Details

Adds support for observing simple string and number bracketed fields (eg: ['with spaces'] and [1337]), as well as identifiers (eg: [symbol]). Support for non-computed numbers and strings is also included eg 1337 and 'with spaces'

String and number literals are very straightforward. Bracketed identifiers (to support symbol) deserve a deeper look.

Support for booleans [true] [false] is not included since it wasn't in the original issue, but it could be desirable for completeness and consistency, although it's hard for me to see why one would do that.

Closes #4547

## Does this pull request introduce a breaking change?

-   💔 Yes, it does introduce a breaking change.

It can be a breaking change if a bracketed identifier, string or number property is expected to not be observable, but I doubt anyone would be relying on this behavior, especially since most linters will warn/error about it.

## Does this pull request introduce an observable change?

-   🔬 Yes, it does include an observable change.

Simple bracketed identifier, strings and number properties will now be observed.

### CODE REVIEW

File: packages/@lwc/integration-karma/test/component/observed-fields/index.spec.js [hunk 1]
```diff
@@ -133,6 +135,57 @@ describe('observed-fields', () => {
...
+        const elm = createElement('x-computed', { is: Computed });
+        document.body.appendChild(elm);
+
+        expect(elm.getValue(`${Computed.prefix}Field`)).toBe('prefixed!');
+        elm.setValue(`${Computed.prefix}Field`, 'mutated');
+
+        return Promise.resolve().then(() => {
+            expect(elm.shadowRoot.querySelector('.prefixed').textContent).toBe('mutated');
+        });
+    });
```
Comment (cardoso) [inline packages/@lwc/integration-karma/test/component/observed-fields/index.spec.js:187]: I included template literals for discussion purposes. It's not in the original issue and should probably be removed before merging.

Comment (wjhsf) [reply packages/@lwc/integration-karma/test/component/observed-fields/index.spec.js:212]: The trouble with template literals, and all computed expressions, is that they can change.

```js
class extends LightningElement {
  [`${Date.now()}`] = 'now'
}
```

In the compiled output, we currently just copy the content of the computed property, i.e. we have `` `${Date.now()}` `` twice. Which means that it's entirely possible that the timestamp in the class definition is a millisecond off from the timestamp in the call to `_registerDecorators`. That's definitely no good!

As described in https://github.com/salesforce/lwc/issues/3956#issuecomment-1901089168, to use computed properties in a more predictable manner, we'll likely need to compute the properties once and store the result. But, like you said, that'll definitely be outside this PR.

Comment (wjhsf) [reply packages/@lwc/integration-karma/test/component/observed-fields/index.spec.js:212]: Also, fwiw, [TypeScript doesn't like](https://www.typescriptlang.org/play/?#code/MYewdgzgLgBBCeBbARiANjAvDAyk1aAFAEQAWAhhcQJQDcAUKJLAAoBKAogGICSAGlhjEADgCcApgDMAlgA9iDJtBhipcrtPFoAJoIAGAEgDe7bvwC+Grdr31xs4SFGxtU8gFc0sYGnIQIMAAq4spG9DAwANrE5MQAuoIADAwRkQCMAMwZAOwJ2MnhUQgo6HlCxQQAhAqFkaoyslY6ZcQgANbk8DWphiacvHyWmjoATHotYCCw7Z3dUXrI7rBQpNIBazAz8OOCAOQAju7Som3wuwzm9EA) computed template literals directly in computed props.

Comment (cardoso) [reply packages/@lwc/integration-karma/test/component/observed-fields/index.spec.js]: Thanks! I thought of that exact case, but still wanted to confirm

**Comment by cardoso** [comment]
I removed the obvserving of template literal computed fields, so this could be ready to merge if everything else is right.

**Comment by cardoso** [comment]
One thing that might be missing here is booleans [true] [false]. No idea why anyone would do that, but it could be included for the sake of completion and consistency.

**Comment by nolanlawson** [comment]
Thanks a lot for this PR! A few quick comments:

1. Since this is an observable change, I think we should put this behind [API versioning](https://developer.salesforce.com/blogs/2024/01/introducing-component-level-api-versioning-for-lwc). It is possible that existing components would be broken if they relied on these properties _not_ being reactive. Also, it's a good carrot for folks to upgrade. This means adding a new feature to [`api-version.ts`](https://github.com/salesforce/lwc/blob/485b715ba3d765d550319493adea057ac1983749/packages/%40lwc/shared/src/api-version.ts#L125-L148) and, for the fixture tests, adding a `config.json` file with `{ "apiVersion": 63 }` to enable the functionality.
2. I agree that booleans are a super-duper edge case and not worth supporting. Even numbers are kinda debatable – only `"strings like this"` and symbols are (I think) really obvious candidates for reactivity.
3. At runtime, we will probably want to restrict this to only expected types, so that people don't use computed properties as a backdoor to allowing every type under the sun. (Not sure how feasible it would be to restrict at the compiler level.)

@ekashida @jye-sf @wjhsf What do you think about supporting numbers?

**Comment by wjhsf** [comment]
> What do you think about supporting numbers?

Any string, numeric, or symbol is a valid property key. If we support strings as property keys (computed and non-computed), then we should also support numeric literals.

I don't think we need to support boolean (and regex and null) literals in computed properties until somebody complains that they actually want to do that.

**Comment by nolanlawson** [comment]
The argument against numbers from my POV is mostly just that it "feels" weird, since it's typically only used in collections (e.g. arrays/HTMLCollections), and I can't imagine why you'd want to use that in an LWC component.

I could also potentially see some issues if we have a reactive property like `1337` which happens to conflict with the string property `"1337"`. In `ReactiveObserver` we use a plain `Object.create(null)` object to track reactive properties, so we'd have to test how the keys might conflict:

https://github.com/salesforce/lwc/blob/13aa150895695ee1781fe24b1e64f13233cd8a88/packages/%40lwc/engine-core/src/libs/mutation-tracker/index.ts#L62

**Comment by wjhsf** [comment]
> I could also potentially see some issues if we have a reactive property like 1337 which happens to conflict with the string property "1337".
Do we currently have handling for duplicate string/identifiers? `{ foo: 1, "foo": 2, foo: 3, "foo": 4 }`

**Comment by nolanlawson** [comment]
It [looks like](https://stackblitz.com/edit/salesforce-lwc-1jtoxj?file=src%2Fmodules%2Fx%2Fapp%2Fapp.js,src%2Fmodules%2Fx%2Fapp%2Fapp.html&title=LWC%20playground) for `@track`ed objects, we already conflate the number property `1337` and the string property `"1337"`. Either are reactive.

I guess maybe this is fine because there's no difference in JavaScript between string keys and number keys in plain objects. I'm just wondering whether we really need to support this at the top-level in components. Even if you want your key to be `[1337]`, it's a lot clearer to use `["1337"]` since this is how JavaScript is going to treat it anyway.

**Comment by nolanlawson** [comment]
I thought about this PR a bit more. It occurred to me that we still don't support decorators (`@api`, `@track`, `@wire`) on these computed properties. So it's not really a full solution.

I feel like it might make more sense to just document this as a limitation of the framework, and then focus on fixing it once browsers have native support for decorators. At that point, it would be easy to support these decorators at the syntax level (since we wouldn't need our own Babel plugin), and we could potentially get rid of `registerDecorators` altogether which would resolve the reactivity issue.

**Comment by cardoso** [comment]
I'm closing this one for now then @nolanlawson . Feel free to reopen if things change.
